apt-hook/json-hook.cc

```diff
@@ -0,0 +1,302 @@
+    if (counts.esm_apps > 0) {
+        std::stringstream ss;
+        ss << counts.esm_apps << " esm-apps ";
+        if (!wrote_security) {
+            ss << "security ";
+        }
+        ss << "update";
+        if (counts.esm_apps != 1) {
+            ss << "s";
+        }
+        count_msgs.push_back(ss.str());
+    }
+
+    for (uint i = 0; i < count_msgs.size(); i++) {
+        if (count_msgs.size() == 3) {
+            if (i == 1) {
+                std::cout << ", ";
+            } else if (i == 2) {
+                std::cout << ", and ";
+            }
+        } else if (count_msgs.size() == 2) {
+            if (i == 1) {
+                std::cout << " and ";
+            }
+        }
+        std::cout << count_msgs[i];
+    }
+    std::cout << std::endl;
+}
+
+int main(int argc, char *argv[])
+{
+    (void) argc;
+    (void) argv;
+
+    char *fd_c_str = getenv("APT_HOOK_SOCKET");
+    if (fd_c_str == NULL) {
+       return 0;
+    }
+    std::string fd_str(fd_c_str);
+    if (fd_str == "") {
+       return 0;
+    }
+    int fd = stoi(fd_str);
+    __gnu_cxx::stdio_filebuf<char> apt_socket_in(fd, std::ios::in);
+    __gnu_cxx::stdio_filebuf<char> apt_socket_out(fd, std::ios::out);
+    std::istream socket_in(&apt_socket_in);
+    std::ostream socket_out(&apt_socket_out);
+
+
+    bool success = false;
+
+    // Read hello message, verify version, and get jsonrpc id
+    jsonrpc_request hello_req;
+    success = read_jsonrpc_request(socket_in, hello_req);
+    if (!success) {
+        return 0;
+    }
+    if (hello_req.method != "org.debian.apt.hooks.hello" || hello_req.notification) {
+        return 0;
+    }
+
+    json_object *hello_req_versions;
+    success = json_object_object_get_ex(hello_req.params, "versions", &hello_req_versions);
+    if (!success) {
+        return 0;
+    }
+    int64_t hello_req_versions_length = json_object_array_length(hello_req_versions);
+
+    bool supports_version_02 = false;
+    for (int64_t i = 0; i < hello_req_versions_length; i++) {
+        std::string version(json_object_get_string(json_object_array_get_idx(hello_req_versions, i)));
+        if (version == "0.2") {
+            supports_version_02 = true;
+            break;
+        }
+    }
+    if (!supports_version_02) {
+        return 0;
+    }
+
+    // Write hello response with jsonrpc id
+    socket_out << "{\"jsonrpc\":\"2.0\",\"id\":" << hello_req.id << ",\"result\":{\"version\":\"0.2\"}\n\n";
+    socket_out.flush();
+
+    json_object_put(hello_req.root_msg);
+
+
+    jsonrpc_request stats_req;
+    success = read_jsonrpc_request(socket_in, stats_req);
+    if (!success) {
+        return 0;
+    }
+    if (stats_req.method == "org.debian.apt.hooks.install.statistics") {
+        security_package_counts counts;
+        success = count_security_packages_from_apt_stats_json(stats_req.params, counts);
+        if (success) {
+            print_counts(counts);
+        }
+    }
+    json_object_put(stats_req.root_msg);
```

Comment:
Is it possible that `stats_req.root_msg` is NULL pointer during this call ?

Reply:
We don't `json_object_put` anything during the processing, so I don't think so.